daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/layers/LayerDriver.scala

```diff
@@ -0,0 +1,338 @@
+      }
+    }
+
+    override def close(): Unit = {
+      try {
+        stream.close()
+        inner.close()
+      } catch {
+        case e: Exception =>
+          handleOutputStreamException(e)
+      }
+    }
+
+    private def handleOutputStreamException(exception: Exception) = {
+      exception match {
+        case u: UnsuppressableException =>
+          throw u
+        case rsde: RuntimeSchemaDefinitionError =>
+          throw rsde
+        case pe: ProcessingError =>
+          throw pe
+        case re: RuntimeException =>
+          throw re
+        case e: Exception =>
+          layer.getLayerRuntime.processingError(new LayerUnexpectedException(e))
+      }
+    }
+
+  }
+}
+
+/**
+ * Turns Daffodil's bits-capable InputSourceDataInputStream objects into ordinary
+ * java InputStream API for byte-oriented reads.
+ *
+ * @param s the Daffodil InputSourceDataInputStream
+ * @return a java.io.InputStream which when read from, delegates to the InputSourceDataInputStream
+ */
+class JavaIOInputStream(s: InputSourceDataInputStream, layer: Layer) extends InputStream {
+  private def finfo = layer.getLayerRuntime.finfo
+  private lazy val id = Misc.getNameFromClass(this)
+
+  override def read(): Int = {
+    if (!s.isDefinedForLength(8)) -1
+    else {
+      val ul = s.getUnsignedLong(8, finfo)
+      val byte: Int = ul.toInt & 0xff
+      byte
+    }
+  }
+
+  override def available(): Int = 0
+
+  override def close(): Unit = {
+    // do nothing. Important. We do not want the close to cascade beyond here.
+  }
+
+  /**
+   * @param readlimit Ignored. The limits of daffodil's input system are specified
+   *                  elsewhere. See BucketingInputSource in the daffodil-io module.
+   */
+  override def mark(readlimit: Int): Unit = {
+    maybeSavedMark = One(s.mark(id))
+  }
+
+  private var maybeSavedMark: Maybe[Mark] = Nope
+
+  override def reset(): Unit = {
+    Assert.usage(maybeSavedMark.isDefined)
+    s.reset(maybeSavedMark.get)
+    maybeSavedMark = Nope
+  }
+
+  override def markSupported() = true
+}
+
+/**
+ * Turns a Daffodil bits-capable DataOutputStream into an ordinary java.io.OutputStream.
+ *
+ * @param dos   The DataOutputStream to write the data to.
+ */
+class JavaIOOutputStream(dos: DataOutputStream, layer: Layer) extends OutputStream {
+
+  private def lr = layer.getLayerRuntime
+  private def lvr = lr.layerVarsRuntime
+  private def finfo = lr.finfo
+
+  private var closed = false
+
+  private var nBytes = 0
+
+  override def write(b: Int): Unit = {
+    val wasWritten = dos.putLong(b, 8, finfo)
+    if (wasWritten) nBytes += 1
+  }
+
+  override def close(): Unit = {
+    if (!closed) {
+      dos.setFinished(finfo)
+      //
+      // This is where we can reliably wrap-up the layer processing.
+      //
+      try {
+        lvr.callGettersToPopulateResultVars(layer)
+      } catch {
+        case ite: InvocationTargetException =>
+          layer.processingError(ite.getCause)
```

Comment:
surprising again. There are definitely test that throw from within the getter in write/unparse tests.